q2_diversity/_alpha/_visualizer.py

```diff
@@ -193,3 +201,147 @@ def alpha_correlation(output_dir: str,
+
+    rows = feature_table.ids(axis='sample')
+    cols = pd.MultiIndex.from_product([list(depth_range), list(iter_range)],
+                                      names=['depth', 'iter'])
+    data = {k: pd.DataFrame(np.NaN, index=rows, columns=cols)
+            for k in metrics}
+
+    for d, i in itertools.product(depth_range, iter_range):
+        rt = rarefy(feature_table, d)
+        for m in metrics:
+            if m in phylogenetic_metrics():
+                vector = alpha_phylogenetic(table=rt, metric=m,
+                                            phylogeny=phylogeny)
+            else:
+                vector = alpha(table=rt, metric=m)
+            data[m][(d, i)] = vector
+    return data
+
+
+def alpha_rarefaction(output_dir: str, table: biom.Table, max_depth: int,
+                      phylogeny: skbio.TreeNode=None, metric: str=None,
+                      metadata: qiime2.Metadata=None, min_depth: int=1,
+                      steps: int=10, iterations: int=10) -> None:
+    if steps < 2:
+        raise ValueError('Provided steps of %d must greater than 1.' % steps)
+    if iterations < 1:
+        raise ValueError('Provided iterations of %d must greater than 0.'
+                         % steps)
+    if metric is None:
+        metrics = ['observed_otus', 'shannon']
+        if phylogeny is not None:
+            metrics.append('faith_pd')
+    else:
+        if metric == 'faith_pd' and phylogeny is None:
+            raise ValueError('Phylogenetic metric %s was requested but '
+                             'phylogeny was not provided.' % metric)
+        metrics = [metric]
+
+    if max_depth <= min_depth:
+        raise ValueError('Provided max_depth of %d must be greater than '
+                         'provided min_depth of %d.' % (max_depth, min_depth))
+    max_frequency = max(table.sum(axis='sample'))
+    if max_frequency < max_depth:
+        raise ValueError('Provided max_depth of %d is greater than '
+                         'the maximum sample total frequency of the '
+                         'feature_table (%d).' % (max_depth, max_frequency))
+    filenames, categories = [], []
+    data = _compute_rarefaction_data(table, min_depth, max_depth,
+                                     steps, iterations, phylogeny, metrics)
+    for (m, data) in data.items():
+        metric_name = quote(m)
+        filename = '%s.csv' % metric_name
+
+        if metadata is None:
+            n_df = _compute_summary(data, 'sample-id')
+            jsonp_filename = '%s.jsonp' % metric_name
+            _beta_rarefaction_jsonp(output_dir, jsonp_filename, metric_name,
+                                    n_df, '')
+            filenames.append(jsonp_filename)
+        else:
+            metadata_df = metadata.to_dataframe()
+            metadata_df.columns = pd.MultiIndex.from_tuples(
+                [(c, '') for c in metadata_df.columns])
+            merged = data.join(metadata_df, how='left')
```

Comment:
👍 

> consistent with other QIIME 2 actions

`alpha-correlation` and `alpha-group-significance` both perform inner joins when joining metadata, taking the intersection of samples in the table and metadata. Could we do that here instead of raising an error?

Reply:
discussed offline, will raise an error!